src/gui/sdlmain.cpp

```diff
@@ -2899,27 +2877,57 @@ static SDL_Point window_bounds_from_label(const std::string& pref,
 	const int percent = [&] {
 		if (pref.starts_with('s')) {
 			return SmallPercent;
+
 		} else if (pref.starts_with('m') || pref == "default" ||
 		           pref.empty()) {
 			return MediumPercent;
+
 		} else if (pref.starts_with('l')) {
 			return LargePercent;
+
 		} else if (pref == "desktop") {
 			return 100;
+
 		} else {
-			LOG_WARNING("DISPLAY: Invalid 'windowresolution' setting: '%s', "
-			            "using 'default'",
-			            pref.c_str());
+			LOG_WARNING(
+			        "DISPLAY: Invalid 'windowresolution' setting: '%s', "
+			        "using 'default'",
+			        pref.c_str());
+
+			set_section_property_value("sdl", "windowresolution", "default");
 			return MediumPercent;
 		}
 	}();
 
+	const auto desktop = get_desktop_size();
+
 	const int w = ceil_sdivide(desktop.w * percent, 100);
 	const int h = ceil_sdivide(desktop.h * percent, 100);
 
 	return {w, h};
 }
 
+static SDL_Point parse_window_resolution_from_conf(const std::string &pref)
+{
+	int w = 0;
+	int h = 0;
+	const bool was_parsed = sscanf(pref.c_str(), "%dx%d", &w, &h) == 2;
+
+	const bool is_valid = (w >= FallbackWindowSize.x &&
+	                       h >= FallbackWindowSize.y);
+
+	if (was_parsed && is_valid) {
+		maybe_limit_requested_resolution(w, h, "window");
+		return {w, h};
+	}
```

Comment:
I think we should clamp here if the user specifies a size that is too small rather than falling through to the percent-based parser.  `FallbackWindowSize` should be... well, the fallback in that case :smile:

Reply:
Well, I don't really want to get into the deep of this by changing it too much... I disagree with the window size defaulting in general in sdlmain, so I'd re-think it from scratch.

The options here are for now:

1. Do nothing (that's bad for reasons I explained in the top description)
2. Default to `auto` which does the percentage thing (ok in my book)
3. Try to clip it (dunno how that would change the overall behaviour in surprising ways, to be inconsistent with all the other window-size determination paths; `auto` is at least something we already have)
4.

Reply:
Well hold on you're changing behavior here.  This is the calling code:

```cpp
	sdl.use_exact_window_resolution = pref.find('x') != std::string::npos;
	if (sdl.use_exact_window_resolution) {
		coarse_size = parse_window_resolution_from_conf(pref);
	} else {
		coarse_size = window_bounds_from_label(pref);
	}
```

It's already doing the check whether it should be parsing for a fixed resolution or a percentage based value.  Then you're falling back to percent based anyway if something went wrong without updating `sdl.use_exact_window_resolution`

Before your change, `parse_window_resolution_from_conf` returned `FallbackWindowSize` if something went wrong.

Reply:
Yes, and I think my changed behaviour is an improvement, but feel free to disagree 😄 

I mentioned this in the top description:

> - Invalid `windowresolution` settings now reset the default window size, not set it to 640x480 pixels (which is passable on non-high DPI displays, but on my Retina MacBook this resulted in a tiny window).

So on a high-DPI screen, if I start up Staging, I get the default window size. So far so good.

Then I go `windowresolution = x` (you need to enable changing `windowresolution` at runtime for that first in the code). Well, that sets it to `FallbackWindowSize`, which is 640x480. But that's in pixels. So now I'm not getting the "default size" I got at startup, but this tiny postage stamp as you can see below 😄 

To me, that's quite broken—defaulting should be a bit more advanced, and at least take high-DPI into account even if we use a hardcoded fallback window size. But that's a more complicated, so I thought, ok, let's fix it by doing what "auto" does. With my fix, `windowresolution = x` will get you the same window size you get at startup without specifying `windowresolution` explicitly, not the postage stamp...

I can revert this whole thing if you really don't like it as I'm not going into this whole window size defaulting rabbit hole. Just wanted to do a small incremental improvement, that's all 😄 

<img width="1728" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/cf91b4e1-36dd-4c6a-81d8-159175c46d65">

<img width="1728" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/2b0a09d9-536f-47c4-9753-9467dcd85c6c">

Reply:
I think it is an improvement but you've stuck it in the wrong place.  Have a look at the calling code in `setup_window_sizes_from_conf`.  `sdl.use_exact_window_resolution` really shouldn't be a global variable.  It's only used in that one function.

You're not updating that variable though and there's more logic that follows that modifies window size:

```cpp
// Save the coarse bounds in the SDL struct for future sizing events
sdl.desktop.requested_window_bounds = {coarse_size.x, coarse_size.y};

// Refine the coarse resolution and save it in the SDL struct.
auto refined_size = coarse_size;
if (sdl.use_exact_window_resolution) {
	refined_size = clamp_to_minimum_window_dimensions(coarse_size);
} else {
	refined_size = refine_window_size(coarse_size,
	                                  wants_aspect_ratio_correction);
}
```

Reply:
> It's already doing the check whether it should be parsing for a fixed resolution or a percentage based value. Then you're falling back to percent based anyway if something went wrong without updating sdl.use_exact_window_resolution

Ok got it, that's not very indeed. I've simplified the logic so the decision whether to use the percentage based or exact dimensions is decided at that top level.

> I think it is an improvement but you've stuck it in the wrong place. Have a look at the calling code in setup_window_sizes_from_conf. sdl.use_exact_window_resolution really shouldn't be a global variable. It's only used in that one function.

Right, turned that into a local var. sdlmain is full with cruft like that 😱